modules/exploits/windows/local/run_as.rb

```diff
@@ -71,11 +71,10 @@ def exploit
       command_line = nil
       windir = get_env('windir')
 
-      # Select path of executable to run depending the architecture
-      if sysinfo['Architecture'] == ARCH_X64 && session.arch == ARCH_X64
+      application_name = "#{windir}\\System32\\notepad.exe"
+      # On x64 systems, select the correct executable path to run if it is a x86 session
+      if sysinfo['Architecture'] == ARCH_X64 && session.arch == ARCH_X86
         application_name = "#{windir}\\SysWOW64\\notepad.exe"
```

Comment:
So using the railgun methods, I'm pretty sure that the session (`session.arch`), target process architecture and payload architecture (`payload.arch.first`) must all be the same. IIRC and from my brief testing attempts, even if the target arch and payload arch match, but the session is different, the railgun methods to inject the shellcode will fail. I'd suggest adding a check here to ensure that everything matches and fail with bad config otherwise.

With that assertion met, you can then actually use the `#get_notepad_pathname` method from the `Msf::Post::Windows::Process` mixin.

Combined those changes would look like:

```suggestion
      fail_with(Failure::BadConfig, 'The payload architecture must match the current session architecture.') unless session.arch == payload.arch.first
      application_name = get_notepad_pathname(session.arch, get_env('windir'), payload.arch.first)
```